### ISSUE TITLE
The server swallows error logs in a case of async function

### ISSUE THREAD

**Issue by kharandziuk** [fastify/fastify#1747]
## 🐛 Bug Report

The server swallows error logs in a case of async function. 

## To Reproduce

bellow 

```js
const supertest = require('supertest')
const { expect } = require('chai')

const Fastify = require('fastify')
let fastify = Fastify()

fastify.get('/normal', (request, reply) => {
  a.send('kapa')
  reply.send({ hello: 'world' })
})

fastify.get('/', async (request, reply) => {
  a.send('kapa')
  return ({ hello: 'world' })
})


describe('server', () => {
  beforeEach(() => {
    return fastify.ready()
  })

  it('error with async', async () => {
    const response = await supertest(fastify.server)
      .get('/')
    expect(response.statusCode).eql(200)
    expect(response.body).eql({ hello: 'world' })
  })

  it('answers hello world', async () => {
    const response = await supertest(fastify.server)
      .get('/normal')
    expect(response.statusCode).eql(200)
    expect(response.body).eql({ hello: 'world' })
  })

})
```

## Expected behavior

I expected that the error logs should be the same but, they are different

```js
  1) server
       error with async:

      AssertionError: expected 500 to deeply equal 200
      + expected - actual

      -500
      +200

      at Context.<anonymous> (test/integration.spec.js:27:33)
      at processTicksAndRejections (internal/process/task_queues.js:85:5)

  2) server
       answers hello world:
     Uncaught ReferenceError: a is not defined
      at Object.<anonymous> (test/integration.spec.js:9:3)
      at preHandlerCallback (node_modules/fastify/lib/handleRequest.js:112:30)
      at preValidationCallback (node_modules/fastify/lib/handleRequest.js:101:5)
      at handler (node_modules/fastify/lib/handleRequest.js:70:5)
      at handleRequest (node_modules/fastify/lib/handleRequest.js:19:5)
      at onRunMiddlewares (node_modules/fastify/lib/middleware.js:22:5)
      at middlewareCallback (node_modules/fastify/lib/route.js:356:5)
      at Object.routeHandler [as handler] (node_modules/fastify/lib/route.js:334:7)
      at Router.lookup (node_modules/find-my-way/index.js:338:14)
      at parserOnIncoming (_http_server.js:713:12)
      at HTTPParser.parserOnHeadersComplete (_http_common.js:116:17)
```

**Comment by mcollina** [comment]
These are not really a _bug_ per se, but a byproduct of how error handling works in Fastify. The error is catched, and we return a 500 instead. We might want to rethrow certain type of errors and crash the process (`RefererenceError` and others). @fastify/fastify wdyt?

As a stopgap solution, you might want to set the log level to `'error'` or less to get the stacktrace in the log.

**Comment by delvedor** [comment]
We already discussed this in #850 and #1078. I fear that decide which error to catch and which not might be tricky and introduce some weird behavior.
IMHO the user should decide what to do and use a custom error handler to specify the behavior.

**Comment by Eomm** [comment]
> the user should decide what to do and use a custom error handler to specify the behavior.

I agree; I think users should manage the errors "by design" and the error management of a promise and a handler execution are different.

Do @kharandziuk think the [Errors.md](https://github.com/fastify/fastify/blob/master/docs/Errors.md#error-handling) doesn't explain this behaviour well?

**Comment by kharandziuk** [comment]
>error management of a promise and a handler execution are different.
the whole point of `async/await` to make it look the same, right?

>doesn't explain this behaviour well?
It kinda describes it, but the actual behavior is cryptic. The problem: in my case, I just changed a function to `async` function and in some point realized that I don't see error stacks anymore. And I noticed it at the moment when the error appears, so It was hard to figure out the reason.
Also, now I need to set up two separate ways to handle 500 errors, right?

also, the documentation says:
>Given the fact that it is not possible to process all uncaught errors in a sensible way, the best way to >deal with them at the moment is to crash.
I believe it somehow brakes this idea

>`supertest(fastify.server)`
The reason why I write the test in such a manner because  I want to see the stack traces. Now I can't do that.


Also, there is another weird dimension of this discussion:
```
fastify.get('/async/promise', async (request, reply) => {
  throw { statusCode: 400, hello: 'world' } // it doesn't log, BTW
})

but not:
fastify.get('/without-async', (request, reply) => {
  throw { statusCode: 400, hello: 'world' } // it will cause a crash
})
```

which makes everything even more cryptic and turns `throw` in a kinda goto.

**Comment by Eomm** [comment]
I think yes because the following use case cannot be caught without creating a sandbox.

```
// it cannot be intercepted by try/catch
fastify.get('/callback-hell', (request, reply) => {
  setTimeout(() => { // whetever DB calls
    a.send('kapa')
    reply.send({ hello: 'world' })
  }, 1000)
})
```

Without changing the API, the only solution that I know for this example is to add:

```
process.on('uncaughtException', (err) => {
  console.log('an error occurred')
})
```

that is really dangerous to me catching all these errors in a node process.

Even Node.js manage the promise rejection in a different manner because a rejected Promis doesn't let the process crash (unless you use the `--unhandled-rejections` flag).

**Comment by kharandziuk** [comment]
you don't understand my question.

```
fastify.get('/callback-hell', (request, reply) => {
  setTimeout(() => { // whetever DB calls
    a.send('kapa')
    reply.send({ hello: 'world' })
  }, 1000)
})
```

this case works exactly as I expect: it just fails with the propper error
```
       error with async -- 2:
     Uncaught ReferenceError: a is not defined
      at Timeout._onTimeout (test/integration.spec.js:15:7)
      at listOnTimeout (internal/timers.js:531:17)
      at processTimers (internal/timers.js:475:7)
```

the thing which I'm asking: How can disable default handling error for async functions?

**Comment by jsumners** [comment]
> the whole point of async/await to make it look the same, right?

No. The point is to provide syntactic sugar around the combination of promises and generators. See https://web.archive.org/web/20190405035312/https://www.promisejs.org/generators/

Do not mix `async function () {}` with callbacks. If you do, then the only viable method is:

```js
function foo (cb) {
  invokeAnAsyncFunction('data')
    .then(result => cb(result))
    .catch(err => cb(null, err))
}
```

**Comment by delvedor** [comment]
> the thing which I'm asking: How can disable default handling error for async functions?

@kharandziuk you can't, but if you use the [`setErrorHandler`](https://www.fastify.io/docs/latest/Server/#seterrorhandler) API you can override our default behavior.

```js
fastify.setErrorHandler(function (error, request, reply) {
  throw error
})
```

**Comment by delvedor** [comment]
Closing this one, feel free to reopen if you feel there is still something we need to discuss.

**Comment by TiredFalcon** [comment]
I was trying to figure out why my errors were not being logged when I found this issue. Similarly, I'm also using `async` for all my plugins and routes.

The [official documentation](https://www.fastify.io/docs/v3.1.x/Server/#seterrorhandler) says that fastify has a default error handler that does this:

```javascript
var statusCode = error.statusCode
if (statusCode >= 500) {
  log.error(error)
} else if (statusCode >= 400) {
  log.info(error)
} else {
  log.error(error)
}
```

So I would expect any error to at least be logged, but no error is being logged.

I then tried using `fastify.setErrorHandler` to log the errors myself, but it didn't work. I called `setErrorHandler` "globally" in the main server file, but I just now discovered that I need to call `setErrorHandler` on _ALL_ my plugins separately. Doing that works.
But I'm using a custom error handler just because I want to print errors, which the default error handler should anyway already do, according to the docs.

So my questions are:
1. Why does the default error handler not log any errors, even if according to the docs it should?
1. Is there a cleaner way to set the same error handler for all my plugins than having to do it in every file?
